src/lib/requests/requests.js

```diff
@@ -99,13 +89,20 @@ class Requests {
 
             return request({ ...reqOpts, agent: this.agent })
               .then(throwOrJson)
-              .catch((e) => {
-                  this.logger.push({ e }).log(`Error attempting HTTP ${method}`);
-                  throw e;
+              .catch((error) => {
+                  this.logger.push({ error }).log(`Error attempting HTTP ${method}`);
+
+                  const { statusCode } = error?.getData?.().res || error || {};
+                  if ([401, 403].includes(statusCode)) {
+                      this.logger.push({ error }).log(`Retrying login due to error statusCode: ${statusCode}`);
+                      const JWT = new JWTSingleton();
+                      return JWT.login().then(() => error).catch(bail);
```

Comment:
Could you please, clarify one thing: 
`...JWT.login().then(() => error)`
what for we return `error` from the promise in case of successful _login()_ call ?

Reply:
Returning the error allow the retry to happen. It looks weird, but this is how retry is working.

Reply:
Sorry, but I still can't get this approach with returning error

1. based on the previous impl., we didn't return an error, but retry worked, right?

2. In previous impl. we did retry only in case `[401, 403].includes(statusCode)`. But now we return error in both cases (line 102 and 99). 

so based on this:
> ...error allow the retry to happen

does it mean we'll retry always (not only in case  `[401, 403].includes(statusCode)`)? 

3. in docs I don't the example with returning an error:
![image](https://github.com/user-attachments/assets/f52614bf-39e8-4b93-8f86-0fe0ef34964c)
could you, please, share a link with this apporach?

Reply:
> based on the previous impl., we didn't return an error, but retry worked, right?

No, the previous approach did exactly the same - see old line 104.
The code from the async-retry example call can throw during `await fetch()`, it is just not done explicitly. The whole point of async-retry is to retry on errors.

Returning an error when login is successful means there is a reason to retry. If the login is not successful, the `bail()` is called to stop retrying.

Reply:
just more questions ))
> Returning an error when login is successful means there is a reason to retry...

but if login is successful, why we need to retry?

Reply:
Throwing and returning is the same in .then() and .catch()

If login is successful we need to retry the original request. Same was happening before my change, with the difference that if login failed the whole process crashed.

Reply:
> Throwing and returning an error is the same in .then() and .catch()

Sorry, but I have to disagree with you

here's a code (return _err_ in the 1st catch):
```
Promise.resolve()
  .then(() => {
    throw new Error('XXX')
  })
  .catch(err => {
    return err
  })
  .catch(err => {
    console.error('ERROR!!!!!!!!!!!!!', err)
  })
```
output (no log):
```
/home/eugen/.nvm/versions/node/v18.17.1/bin/node /home/eugen/projects/ML/central-services-logger/src/aStorage.js

Process finished with exit code 0
```

and here's we return _err_ in the 1st catch:
```
Promise.resolve()
  .then(() => {
    throw new Error('XXX')
  })
  .catch(err => {
    throw err
  })
  .catch(err => {
    console.error('ERROR!!!!!!!!!!!!!', err)
  })
```
output:
```
/home/eugen/.nvm/versions/node/v18.17.1/bin/node /home/eugen/projects/ML/central-services-logger/src/aStorage.js
ERROR!!!!!!!!!!!!! Error: XXX
    at /home/eugen/projects/ML/central-services-logger/src/aStorage.js:56:11

Process finished with exit code 0

```

Reply:
> Throwing and returning an error is the same in .then() and .catch()

the difference is like here:
```
async fn1() {
  try {
     // ... some logic
  } catch (err) {
      return err
  }
}

async fn2() {
  try {
     // ... some logic
  } catch (err) {
      throw err
  }
}
```
`fn1` finishes successfully, and  _returns_ a value (_err_ in this case), whereas `fn2` _throws_ and exception

Reply:
Sorry, I checked again, I need to throw it indeed.